### PR TITLE
feat: m2 mainnet upgrade script

### PR DESCRIPTION
**Pull request by 8sunyuan**
**M2-mainnet LFGGGG**

Script for deploying and doing sanity checks found under `script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol`
Run script locally with 
```
anvil --fork-url $RPC_MAINNET
forge script script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol --rpc-url http://127.0.0.1:8545 --private-key $PRIVATE_KEY --broadcast -vvvv
```

Uses two config json files:
`script/output/mainnet/M1_deployment_mainnet_2023_6_9.json` 
Used for existing mainnet contract addresses (slightly modified to fix json parse errors)

`script/configs/mainnet/M2_mainnet_upgrade.config.json`
Used for expected initalization values. Sanity checks against these config values after calling `_deployImplementationContracts()` and `_simulateUpgrade()`

### CODE REVIEW

File: script/configs/mainnet/M2_mainnet_upgrade.config.json
Comment (ChaoticWalrus) [inline script/configs/mainnet/M2_mainnet_upgrade.config.json:45]: just reviewed all params in this config; LGTM

File: script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol
Comment (ChaoticWalrus) [inline script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol:33]: I guess we could also add the AVSDirectory.delegationManager() to this?

Comment (ChaoticWalrus) [reply script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol:33]: oh, this is already done, incredible.

File: script/output/mainnet/M1_deployment_mainnet_2023_6_9.json
Comment (ChaoticWalrus) [inline script/output/mainnet/M1_deployment_mainnet_2023_6_9.json:6]: should this be non-zero? I get that the `avsDirectory` is set to the zero address because it is not deployed yet.
I would guess that this doesn't matter, just searching for anything, honestly.

**Review by ChaoticWalrus — CHANGES_REQUESTED** [review]
This is looking excellent!
I'm marking this as 'request changes' for the moment, but purely because I want to ensure this gets multiple reviews before it's merged; I couldn't find anything amiss, myself.

File: script/output/mainnet/M1_deployment_mainnet_2023_6_9.json
Comment (8sunyuan) [reply script/output/mainnet/M1_deployment_mainnet_2023_6_9.json:6]: Ah yes I added this because the `_parseDeployedContracts` helper will throw an error when trying to parse it out. It was used for the holesky deployment so I added these 0 addresses here based on the current mainnet state.
The beaconOracle address gets overwritten when `_parseInitialDeploymentParams` is called right after.

File: script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol
Comment (ypatil12) [inline script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol:11]: The scripts LGTM, but when running the command locally (without broadcast), I get `Multiple contracts in the target path. Please specify the contract name with `--tc ContractName`.

**Review by wadealexc — COMMENTED** [review]
Do we still need `script/deploy/mainnet/M2Deploy.s.sol`?

File: script/utils/ExistingDeploymentParser.sol
Comment (wadealexc) [inline script/utils/ExistingDeploymentParser.sol:120]: I see this is added here, but don't see a matching line added to `_parseInitialDeploymentParams`. Is that important?

File: script/configs/mainnet/M2_mainnet_upgrade.config.json
Comment (wadealexc) [inline script/configs/mainnet/M2_mainnet_upgrade.config.json:33]: Is this correct? it looks to be denominated in wei, not gwei

File: script/configs/mainnet/M2_mainnet_upgrade.config.json
Comment (wadealexc) [inline script/configs/mainnet/M2_mainnet_upgrade.config.json:34]: Where are we getting this? Not that it's wrong, but I want a source to double check this.

File: script/configs/mainnet/M2_mainnet_upgrade.config.json
Comment (wadealexc) [inline script/configs/mainnet/M2_mainnet_upgrade.config.json:38]: Where are we getting this? Not that it's wrong, but I want a source to double check this.

File: script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol
Comment (wadealexc) [inline script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol:85]: Where does migrating queued withdrawals come in? Do we do that separately?

File: script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol
Comment (wadealexc) [inline script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol:149]: I don't think this is needed? The deployment parser picks this up from the JSON config

File: script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol
Comment (ChaoticWalrus) [reply script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol:85]: yes, that would be completed separately, post-upgrade

File: script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol
Comment (wadealexc) [inline script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol:223]: I assume these transactions are executed en masse, but in case they aren't (or just to avoid footgun) - let's swap the order of these two transactions and make sure the withdrawal delay is configured before we unpause the contract.

File: script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol
Comment (wadealexc) [inline script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol:244]: Same request here - configure, then unpause.

File: script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol
Comment (ChaoticWalrus) [reply script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol:149]: good point, will resolve

File: script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol
Comment (ChaoticWalrus) [reply script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol:223]: they are all completed as part of a single transaction. I can change the order though

**Review by wadealexc — CHANGES_REQUESTED** [review]
Have left several questions and some requested changes

Should be fine once those are addressed.

I glossed over some of the multisig interactions/queue upgrade test -- will make time to review that stuff more closely when it's time to do the actual upgrade. 

The deploy stuff looks good otherwise.

File: script/configs/mainnet/M2_mainnet_upgrade.config.json
Comment (ChaoticWalrus) [reply script/configs/mainnet/M2_mainnet_upgrade.config.json:38]: The Geth release has this https://github.com/ethereum/go-ethereum/releases/tag/v1.13.12

Comment (8sunyuan) [reply script/configs/mainnet/M2_mainnet_upgrade.config.json:38]: https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/fork.md#configuration
https://beaconcha.in/epoch/269568
Fork epoch is `269568` and I converted to the timestamp

File: script/configs/mainnet/M2_mainnet_upgrade.config.json
Comment (ChaoticWalrus) [reply script/configs/mainnet/M2_mainnet_upgrade.config.json:34]: the phase0 spec has this https://github.com/ethereum/consensus-specs/blob/dev/specs/phase0/beacon-chain.md but it appears to be the 'minimum' time, not the set time.
looking at the explorer https://beaconcha.in/slot/0 I see `1606824023` instead, and there is a blog post confirming it.
so this should be `1606824023`; good catch

File: script/configs/mainnet/M2_mainnet_upgrade.config.json
Comment (8sunyuan) [reply script/configs/mainnet/M2_mainnet_upgrade.config.json:38]: Found the actual config in the specs here https://github.com/ethereum/consensus-specs/blob/68dceb7cda9f7745689fda007d712e6b128f94ac/configs/mainnet.yaml#L52

File: script/configs/mainnet/M2_mainnet_upgrade.config.json
Comment (8sunyuan) [reply script/configs/mainnet/M2_mainnet_upgrade.config.json:34]: https://github.com/ethereum/consensus-specs/blob/68dceb7cda9f7745689fda007d712e6b128f94ac/configs/mainnet.yaml#L52

Comment (ChaoticWalrus) [reply script/configs/mainnet/M2_mainnet_upgrade.config.json]: fixed in https://github.com/Layr-Labs/eigenlayer-contracts/pull/480/commits/02f7aff812fa5e73c6f0fdfe056b622444112797

File: script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol
Comment (ChaoticWalrus) [reply script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol]: fixed in https://github.com/Layr-Labs/eigenlayer-contracts/pull/480/commits/9b7a0c33c94c20f2e5bb8abfacf115e15bf24426

File: script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol
Comment (ChaoticWalrus) [reply script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol]: adjusted the ordering in https://github.com/Layr-Labs/eigenlayer-contracts/pull/480/commits/9b7a0c33c94c20f2e5bb8abfacf115e15bf24426

File: script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol
Comment (ChaoticWalrus) [reply script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol]: adjusted the ordering in https://github.com/Layr-Labs/eigenlayer-contracts/pull/480/commits/9b7a0c33c94c20f2e5bb8abfacf115e15bf24426

File: script/configs/mainnet/M2_mainnet_upgrade.config.json
Comment (8sunyuan) [reply script/configs/mainnet/M2_mainnet_upgrade.config.json]: https://github.com/Layr-Labs/eigenlayer-contracts/pull/480/commits/888d3c6b02a8f0f2261ca7590b0852f3c98cccd0

Comment (8sunyuan) [reply script/configs/mainnet/M2_mainnet_upgrade.config.json]: Added a modulo check in case this value gets overflowed again

File: script/configs/mainnet/M2_mainnet_upgrade.config.json
Comment (wadealexc) [reply script/configs/mainnet/M2_mainnet_upgrade.config.json]: Interesting -- do we know if the holesky value is correct?

File: script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol
Comment (wadealexc) [reply script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol]: thanks, i feel better about this :)

File: script/configs/mainnet/M2_mainnet_upgrade.config.json
Comment (8sunyuan) [reply script/configs/mainnet/M2_mainnet_upgrade.config.json]: https://github.com/eth-clients/holesky
https://holesky.beaconcha.in/epoch/0
I checked the holesky client page initially, seems to align with the explorer

**Review by wadealexc — APPROVED** [review]
Fixes look good!

File: script/configs/mainnet/M2_mainnet_upgrade.config.json
Comment (ChaoticWalrus) [reply script/configs/mainnet/M2_mainnet_upgrade.config.json]: LGTM, thank you

File: script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol
Comment (ChaoticWalrus) [reply script/deploy/mainnet/M2_Mainnet_Upgrade.s.sol]: this was addressed in https://github.com/Layr-Labs/eigenlayer-contracts/pull/480/commits/888d3c6b02a8f0f2261ca7590b0852f3c98cccd0 😄

**Review by ChaoticWalrus — APPROVED** [review]
seems we have agreement this is looking good. going to merge it.

**Pull request merged**
